pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenJWK.java

```diff
@@ -0,0 +1,358 @@
+
+        if (audienceClaim != null && audience == null ) {
+            throw new IllegalArgumentException("Token Audience Claim [" + audienceClaim
+                    + "] configured, but Audience stands for this broker not.");
+        }
+    }
+
+    @Override
+    public String getAuthMethodName() {
+        return TOKEN;
+    }
+
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+        try {
+            // Get Token
+            String token;
+            String role;
+            token = getToken(authData);
+            // Parse Token by validating
+
+
+            role = getPrincipal(authenticateToken(token));
+
+
+            AuthenticationMetrics.authenticateSuccess(getClass().getSimpleName(), getAuthMethodName());
+            return role;
+        } catch (AuthenticationException exception) {
+            AuthenticationMetrics.authenticateFailure(getClass().getSimpleName(), getAuthMethodName(), exception.getMessage());
+            throw exception;
+        }
+    }
+
+    @Override
+    public AuthenticationState newAuthState(AuthData authData, SocketAddress remoteAddress, SSLSession sslSession)
+            throws AuthenticationException {
+        return new TokenAuthenticationState(this, authData, remoteAddress, sslSession);
+    }
+
+    public static String getToken(AuthenticationDataSource authData) throws AuthenticationException {
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            return validateToken(authData.getCommandData());
+        } else if (authData.hasDataFromHttp()) {
+            // Authentication HTTP request. The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                throw new AuthenticationException("Invalid HTTP Authorization header");
+            }
+
+            // Remove prefix
+            String token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+            return validateToken(token);
+        } else {
+            throw new AuthenticationException("No token credentials passed");
+        }
+    }
+
+    private static String validateToken(final String token) throws AuthenticationException {
+        if (StringUtils.isNotBlank(token)) {
+            return token;
+        } else {
+            throw new AuthenticationException("Blank token found");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private DecodedJWT authenticateToken(final String token) throws AuthenticationException {
+
+        DecodedJWT jwt = JWT.decode(token);
+
+        Jwk jwk = null;
+        Algorithm algorithm = null;
+        try {
+            jwk = provider.get(jwt.getKeyId());
```

Comment:
This method relies on a blocking call to the Authorization Server. I think we should improve the interface to allow this call to be asynchronous. Otherwise, we'll block a netty thread.

@lhotari and I noticed this issue earlier today, and I sent an email to the pulsar dev mailing list proposing a solution: https://lists.apache.org/x/thread.html/r6c2522ca62242109758586696261cb1f4b4ce8e94ae593fda6e97b99@%3Cdev.pulsar.apache.org%3E